### PR TITLE
Add explicit t->s transactions with memo field

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/40602

Allows to send from t -> s with adding memo field.
ZCashCreateShieldTransactionTask is reused both for shield all and t->s transactions.
Extracted tx signing to a separate class - ZCashTransactionCompleteManager.

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/common/zcash_utils.cc [hunk 1]
```diff
@@ -11,13 +11,15 @@
 #include <vector>
 
 #include "base/containers/span.h"
+#include "base/debug/stack_trace.h"
 #include "base/numerics/byte_conversions.h"
```
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.cc:15]: ~#include "base/debug/stack_trace.h"~

File: components/brave_wallet/browser/zcash/zcash_transaction.h [hunk 2]
```diff
@@ -139,6 +136,9 @@ class ZCashTransaction {
   std::string to() const { return to_; }
   void set_to(const std::string& to) { to_ = to; }
 
+  std::optional<OrchardMemo> memo() const { return memo_; }
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_transaction.h:139]: const std::optional<OrchardMemo>& memo() const { return memo_; }

File: components/brave_wallet/browser/zcash/zcash_transaction.h [hunk 3]
```diff
@@ -139,6 +136,9 @@ class ZCashTransaction {
...
+  void set_memo(std::optional<OrchardMemo> memo) { memo_ = memo; }
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_transaction.h:140]: void set_memo(const std::optional<OrchardMemo>& memo) { memo_ = memo; }

File: components/brave_wallet/browser/zcash/zcash_transaction.cc [hunk 4]
```diff
@@ -358,6 +274,11 @@ base::Value::Dict ZCashTransaction::ToValue() const {
     outputs_value.Append(output.ToValue());
   }
   // TODO(cypt4): Add orchard part serialization\deserialization
+  auto& orchard_outputs_value =
+      dict.Set("orchard_outputs", base::Value::List())->GetList();
+  for (auto& orchard_output : orchard_part_.outputs) {
+    orchard_outputs_value.Append(orchard_output.ToValue());
+  }
 
   dict.Set("locktime", base::NumberToString(locktime_));
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_transaction.cc:283]: Should memo be there?

File: components/brave_wallet/browser/zcash/zcash_transaction.cc [hunk 5]
```diff
@@ -388,18 +309,36 @@ std::optional<ZCashTransaction> ZCashTransaction::FromValue(
   }
 
   auto* outputs_list = value.FindList("outputs");
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_transaction.cc:311]: Should memo be there?

File: components/brave_wallet/browser/zcash/zcash_transaction.h [hunk 6]
```diff
@@ -166,6 +166,7 @@ class ZCashTransaction {
   uint32_t locktime_ = 0;
   uint32_t expiry_height_ = 0;
   std::string to_;
+  std::optional<OrchardMemo> memo_;
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_transaction.h:169]: should be compared in `operator ==`

File: components/brave_wallet/browser/zcash/zcash_create_transparent_transaction_task.cc [hunk 7]
```diff
@@ -77,12 +81,17 @@ void ZCashCreateTransparentTransactionTask::WorkOnTask() {
...
+  if (!pick_inputs_result) {
     // TODO(cypt4) : switch to IDS_BRAVE_WALLET_INSUFFICIENT_BALANCE when ready
     SetError(l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
     WorkOnTask();
     return;
   }
+  transaction_.set_fee(pick_inputs_result->fee);
+  transaction_.transparent_part().inputs.insert(
+      transaction_.transparent_part().inputs.end(),
+      pick_inputs_result->inputs.begin(), pick_inputs_result->inputs.end());
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_create_transparent_transaction_task.cc:95]: base::Extend

File: components/brave_wallet/common/zcash_utils.h [hunk 8]
```diff
@@ -73,8 +80,12 @@ struct DecodedZCashAddress {
 struct OrchardOutput {
   uint32_t value = 0;
   std::array<std::uint8_t, ::brave_wallet::kOrchardRawBytesSize> addr;
```
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.h:82]: OrchardAddrRawPart here and everywhere

File: components/brave_wallet/common/zcash_utils.cc [hunk 9]
```diff
@@ -370,4 +408,29 @@ std::optional<std::string> GetOrchardUnifiedAddress(
...
+  if (!input) {
+    return std::nullopt;
+  }
+
+  if (input->size() > kOrchardMemoSize) {
+    return std::nullopt;
+  }
+
+  std::array<uint8_t, kOrchardMemoSize> output;
+  output.fill(0);
```
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.cc:422]: std::array<uint8_t, kOrchardMemoSize> output = {}
~output.fill(0);~

File: components/brave_wallet/common/zcash_utils.cc [hunk 10]
```diff
@@ -370,4 +408,29 @@ std::optional<std::string> GetOrchardUnifiedAddress(
       testnet);
 }
 
+std::optional<OrchardMemo> ToOrchardMemo(
+    std::optional<std::vector<uint8_t>> input) {
```
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.cc:412]: const &

File: components/brave_wallet/common/zcash_utils.h [hunk 11]
```diff
@@ -6,12 +6,14 @@
...
 #define BRAVE_COMPONENTS_BRAVE_WALLET_COMMON_ZCASH_UTILS_H_
 
+#include <limits>
 #include <optional>
 #include <string>
 #include <utility>
 #include <vector>
 
 #include "base/containers/span.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
```
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.h:16]: need this include here?

**Review by supermassive — APPROVED** [review]
wallet core lgtm

**Review by Douglashdaniel — APPROVED** [review]
++ Wallet Front-end

**Review by nuo-xu — APPROVED** [review]
iOS++

**Pull request merged**